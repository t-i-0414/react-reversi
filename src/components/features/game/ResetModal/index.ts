export { ResetModalContainer as ResetModal } from './ResetModalContainer';
