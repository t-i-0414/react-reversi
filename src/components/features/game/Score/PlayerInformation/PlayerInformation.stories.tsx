import { Meta, StoryObj } from '@storybook/react';
import PlayerInformation from './PlayerInformation';

const meta: Meta<typeof PlayerInformation> = {
  title: 'features/game/Score/PlayerInformation',
  component: PlayerInformation,
  argTypes: {
    name: {
      control: 'text',
    },
    score: {
      control: 'number',
    },
    pieceColor: {
      control: 'select',
      options: ['black', 'white'],
    },
    nameSide: {
      control: 'select',
      options: ['left', 'right'],
    },
  },
  args: {
    name: 'Player1',
    score: 10,
    pieceColor: 'black',
    nameSide: 'left',
  },
};

export default meta;

type Story = StoryObj<typeof PlayerInformation>;

export const Left: Story = {
  args: {
    name: 'Player1',
    pieceColor: 'black',
    nameSide: 'left',
  },
  render: props => <PlayerInformation {...props} />,
};

export const Right: Story = {
  args: {
    name: 'Player2',
    pieceColor: 'white',
    nameSide: 'right',
  },
  render: props => <PlayerInformation {...props} />,
};
