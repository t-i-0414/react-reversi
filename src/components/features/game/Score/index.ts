export { ScoreContainer as Score } from './ScoreContainer';
