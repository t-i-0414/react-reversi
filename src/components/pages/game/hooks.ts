import { useSelector } from 'react-redux';
import type { Store } from '~/types';
import { PieceColor } from '~/const';
import Utils from '~/utils';

const {
  Game: { getUpdatableSquaresArray },
} = Utils;

const useGame = (): {
  isGameStarted: boolean;
  isGameFinished: boolean;
} => {
  const { isGameStarted, board } = useSelector((store: Store) => store.game);

  const updatableSquaresArrays = board.filter(
    square => getUpdatableSquaresArray(square).length > 0,
  );

  const canPlacePieces = !!updatableSquaresArrays.find(
    square => square.pieceColor === PieceColor.INVISIBLE,
  );

  const isGameFinished = isGameStarted && !canPlacePieces;

  return {
    isGameStarted,
    isGameFinished,
  };
};
export default useGame;
