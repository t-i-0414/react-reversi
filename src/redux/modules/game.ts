import { Dispatch } from 'redux';
import { PlayerMap, PieceColor } from '~/const';
import type { Square, Board, Store, MapValues } from '~/types';

/**
 * action
 */
const UPDATE_GAME_START_FLAG = 'reversi/store/game/update-game-start-flag';
const UPDATE_SIDE_SQUARES_COUNT =
  'reversi/store/game/update-side-squares-count';
const UPDATE_BOARD = 'reversi/store/game/update-board';
const UPDATE_PLAYERS = 'reversi/store/game/update-players';
const CLEAR_STATE = 'reversi/store/game/clear-state';
const DO_NOTHING = 'reversi/store/game/do-nothing';

/**
 * action types
 */
type UpdateGameStartFlag = {
  type: typeof UPDATE_GAME_START_FLAG;
  payload: boolean;
};

type UpdateSideSquaresCount = {
  type: typeof UPDATE_SIDE_SQUARES_COUNT;
  payload: number;
};

type UpdateBoard = {
  type: typeof UPDATE_BOARD;
  payload: Board;
};

type UpdatePlayers = {
  type: typeof UPDATE_PLAYERS;
  payload: Store['game']['players'];
};

type ClearState = {
  type: typeof CLEAR_STATE;
};

type DoNothing = {
  type: typeof DO_NOTHING;
};

type ActionType =
  | UpdateGameStartFlag
  | UpdateSideSquaresCount
  | UpdateBoard
  | UpdatePlayers
  | ClearState
  | DoNothing;

/**
 * initial state
 */
export const initialState: Store['game'] = {
  isGameStarted: false,
  sideSquaresCount: 0,
  board: [],
  players: {
    black: {
      ...PlayerMap.PLAYER_1,
      pieceColor: PieceColor.BLACK,
      score: 0,
      current: true,
    },
    white: {
      ...PlayerMap.PLAYER_2,
      pieceColor: PieceColor.WHITE,
      score: 0,
      current: false,
    },
  },
};

/**
 * reducer
 */
export const gameReducer = (
  state: Store['game'] = initialState,
  action: ActionType = {
    type: 'reversi/store/game/do-nothing',
  },
): Store['game'] => {
  switch (action.type) {
    case UPDATE_GAME_START_FLAG:
      return {
        ...state,
        isGameStarted: action.payload,
      };
    case UPDATE_SIDE_SQUARES_COUNT:
      return {
        ...state,
        sideSquaresCount: action.payload,
      };
    case UPDATE_BOARD:
      return {
        ...state,
        board: action.payload,
      };
    case UPDATE_PLAYERS:
      return {
        ...state,
        players: action.payload,
      };
    case CLEAR_STATE:
      return {
        ...initialState,
      };
    case DO_NOTHING:
      return state;
    default: {
      const _unreachable: never = action;

      return state;
    }
  }
};

export const updateGameStartFlg = (flag: boolean): UpdateGameStartFlag => ({
  type: UPDATE_GAME_START_FLAG,
  payload: flag,
});

export const updateSideSquaresCount = (
  sideSquaresCount: number,
): UpdateSideSquaresCount => ({
  type: UPDATE_SIDE_SQUARES_COUNT,
  payload: sideSquaresCount,
});

export const updateBoard = (stagingBoard: Board): UpdateBoard => ({
  type: UPDATE_BOARD,
  payload: stagingBoard,
});

export const updatePlayers = (
  stagingPlayers: Store['game']['players'],
): UpdatePlayers => ({
  type: UPDATE_PLAYERS,
  payload: stagingPlayers,
});

export const clearState = (): ClearState => ({
  type: CLEAR_STATE,
});

export const initializeBoard =
  (sideSquaresCount: number) =>
  (dispatch: Dispatch): void => {
    const squaresCountAmount: number = sideSquaresCount ** 2;
    const stagingBoard: Board = [];

    for (
      let squareCount = 0;
      squareCount < squaresCountAmount;
      squareCount += 1
    ) {
      const key = squareCount;
      const column = squareCount % sideSquaresCount;
      const row = Math.floor(squareCount / sideSquaresCount);
      let pieceColor: MapValues<typeof PieceColor> = PieceColor.INVISIBLE;

      /**
       * place four stones when the board is initially rendered
       */
      // upprer left square
      if (column === sideSquaresCount / 2 && row === sideSquaresCount / 2 - 1) {
        pieceColor = PieceColor.WHITE;
      }

      // upper right square
      if (
        column === sideSquaresCount / 2 - 1 &&
        row === sideSquaresCount / 2 - 1
      ) {
        pieceColor = PieceColor.BLACK;
      }

      // lower left square
      if (column === sideSquaresCount / 2 && row === sideSquaresCount / 2) {
        pieceColor = PieceColor.BLACK;
      }

      // lower right square
      if (column === sideSquaresCount / 2 - 1 && row === sideSquaresCount / 2) {
        pieceColor = PieceColor.WHITE;
      }

      stagingBoard.push({
        key,
        column,
        row,
        pieceColor,
      });
    }

    dispatch(updateSideSquaresCount(sideSquaresCount));
    dispatch(updateBoard(stagingBoard));
  };

export const updateScore =
  () =>
  (dispatch: Dispatch, getState: () => Store): void => {
    const {
      game: { board, players },
    } = getState();

    const whitePiecesCount = board.reduce(
      (prev, square) => prev + (square.pieceColor === PieceColor.WHITE ? 1 : 0),
      0,
    );

    const blackPiecesCount = board.reduce(
      (prev, square) => prev + (square.pieceColor === PieceColor.BLACK ? 1 : 0),
      0,
    );

    const stagingPlayers = { ...players };

    stagingPlayers.white.score = whitePiecesCount;
    stagingPlayers.black.score = blackPiecesCount;

    dispatch(updatePlayers(stagingPlayers));
  };

export const changeGamesTurn =
  (square: Square, updatableSquaresArray: Square[]) =>
  (dispatch: Dispatch, getState: () => Store): void => {
    const {
      game: { board: stagingBoard, players },
    } = getState();
    const clickedSquare: Square = stagingBoard[square.key];

    const currentPlayerIndex = Object.entries(players).find(
      ([_id, player]) => player.current === true,
    )?.[0] as keyof Store['game']['players'];
    const currentPlayer = players[currentPlayerIndex];

    // change each the value of squares for pieces to be turn over
    updatableSquaresArray.forEach((updatableSquare: Square) => {
      stagingBoard[updatableSquare.key].pieceColor = currentPlayer.pieceColor;
    });

    // change the value of clicked square
    clickedSquare.pieceColor = currentPlayer.pieceColor;

    // switch the current player
    const stagingPlayers = Object.fromEntries(
      Object.entries(players).map(([key, player]) => {
        const stagingPlayer = { ...player };
        stagingPlayer.current = !stagingPlayer.current;

        return [key, stagingPlayer];
      }),
    ) as Store['game']['players'];

    dispatch(updateBoard(stagingBoard));
    dispatch(updatePlayers(stagingPlayers));
    updateScore();
  };
