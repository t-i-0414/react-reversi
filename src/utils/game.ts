import { PieceColor } from '~/const';
import { store } from '~/redux';
import type { Square, Store } from '~/types';

/**
 * Function to get arrays of the square that has piece that can be turn over in each directions from the passed square
 */
const getUpdatableSquaresArray = (passedSquare: Square): Square[] => {
  const { board, sideSquaresCount, players } = store.getState().game;

  const currentPlayerIndex = Object.entries(players).find(
    ([_id, player]) => player.current === true,
  )?.[0] as keyof Store['game']['players'];
  const currentPlayer = players[currentPlayerIndex];

  const eachDirectionSquaresArray: Array<Square[]> = [];

  // Get an array of squares from the passed square to the left
  eachDirectionSquaresArray.push(
    board
      .filter(
        square =>
          !!(
            square.row === passedSquare.row &&
            square.column < passedSquare.column
          ),
      )
      .reverse(), // Reverse to use the passed square as a starting point
  );

  // Get an array of squares from the passed square to the right side
  eachDirectionSquaresArray.push(
    board.filter(
      square =>
        !!(
          square.row === passedSquare.row && square.column > passedSquare.column
        ),
    ),
  );

  // Get an array of squares from the passed square to the top
  eachDirectionSquaresArray.push(
    board
      .filter(
        square =>
          !!(
            square.row < passedSquare.row &&
            square.column === passedSquare.column
          ),
      )
      .reverse(), // Reverse to use the passed square as a starting point
  );

  // Get an array of squares from the passed square to the bottom
  eachDirectionSquaresArray.push(
    board.filter(
      square =>
        !!(
          square.row > passedSquare.row && square.column === passedSquare.column
        ),
    ),
  );

  // Get an array of squares from the passed square to the upper left
  const upperLeftSquaresArray: Square[] = [];
  for (let count = 1; count < sideSquaresCount; count += 1) {
    const squareId: number =
      passedSquare.key - sideSquaresCount * count - count;

    const upperLeftSquare = board[squareId];

    if (
      upperLeftSquare &&
      upperLeftSquare.key < passedSquare.key &&
      upperLeftSquare.column < passedSquare.column &&
      upperLeftSquare.row < passedSquare.row
    ) {
      upperLeftSquaresArray.push(upperLeftSquare);
    }
  }
  eachDirectionSquaresArray.push(upperLeftSquaresArray);

  // Get an array of squares from the passed square to the upper right
  const upperRightSquaresArray: Square[] = [];
  for (let count = 1; count < sideSquaresCount; count += 1) {
    const squareId: number =
      passedSquare.key - sideSquaresCount * count + count;

    const upperRightSquare = board[squareId];

    if (
      upperRightSquare &&
      upperRightSquare.key < passedSquare.key &&
      upperRightSquare.column > passedSquare.column &&
      upperRightSquare.row < passedSquare.row
    ) {
      upperRightSquaresArray.push(upperRightSquare);
    }
  }
  eachDirectionSquaresArray.push(upperRightSquaresArray);

  // Get an array of squares from the passed square to the lower left
  const lowerLeftSquaresArray: Square[] = [];
  for (let count = 1; count < sideSquaresCount; count += 1) {
    const squareId: number =
      passedSquare.key + sideSquaresCount * count - count;

    const lowerLeftSquare = board[squareId];

    if (
      lowerLeftSquare &&
      lowerLeftSquare.key > passedSquare.key &&
      lowerLeftSquare.column < passedSquare.column &&
      lowerLeftSquare.row > passedSquare.row
    ) {
      lowerLeftSquaresArray.push(lowerLeftSquare);
    }
  }
  eachDirectionSquaresArray.push(lowerLeftSquaresArray);

  // Get an array of squares from the passed square to the lower right
  const lowerRightSquaresArray: Square[] = [];
  for (let count = 1; count < sideSquaresCount; count += 1) {
    const squareId: number =
      passedSquare.key + sideSquaresCount * count + count;

    const lowerRightSquare = board[squareId];

    if (
      lowerRightSquare &&
      lowerRightSquare.key > passedSquare.key &&
      lowerRightSquare.column > passedSquare.column &&
      lowerRightSquare.row > passedSquare.row
    ) {
      lowerRightSquaresArray.push(lowerRightSquare);
    }
  }
  eachDirectionSquaresArray.push(lowerRightSquaresArray);

  const updatableSquaresArray: Square[] = [];

  eachDirectionSquaresArray.forEach(squareArray => {
    const emptySquareIndex: number = squareArray.findIndex(
      square => square.pieceColor === PieceColor.INVISIBLE,
    );
    if (emptySquareIndex !== -1) {
      squareArray.splice(emptySquareIndex, squareArray.length);
    }

    const endpointSquareIndex: number = squareArray.findIndex(
      square => square.pieceColor === currentPlayer.pieceColor,
    );
    squareArray.splice(Math.max(0, endpointSquareIndex), squareArray.length);

    updatableSquaresArray.push(...squareArray);
  });

  return updatableSquaresArray;
};

export default { getUpdatableSquaresArray };
