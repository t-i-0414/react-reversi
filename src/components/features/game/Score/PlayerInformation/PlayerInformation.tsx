import React, { memo } from 'react';
import isEqual from 'react-fast-compare';
import styled from 'styled-components';
import { Piece } from '~/components/features/game/Piece';
import { SizeMap } from '~/constants';
import type {
  PieceColor,
  PlayerInformation as TPlayerInformation,
} from '~/domains';

export type Props = TPlayerInformation & {
  nameSide: 'left' | 'right';
};

export const PlayerInformation: React.FC<Props> = ({
  name,
  pieceColor,
  score,
  nameSide,
}) => (
  <Wrapper>
    {nameSide === 'left' && (
      <Player name={name} pieceColor={pieceColor} side={nameSide} />
    )}

    <Score>{score}</Score>

    {nameSide === 'right' && (
      <Player name={name} pieceColor={pieceColor} side={nameSide} />
    )}
  </Wrapper>
);

const Player: React.FC<{
  name: string;
  pieceColor: PieceColor;
  side: 'left' | 'right';
}> = memo(
  ({ name, pieceColor, side }) => (
    <>
      {side === 'left' && (
        <>
          <PlayerName>{name}</PlayerName>
          <Piece pieceColor={pieceColor} size='48px' />
        </>
      )}
      {side === 'right' && (
        <>
          <Piece pieceColor={pieceColor} size='48px' />
          <PlayerName>{name}</PlayerName>
        </>
      )}
    </>
  ),
  isEqual,
);
Player.displayName = 'Player';

const Wrapper = styled.div`
  display: flex;
  gap: 4px;
  align-items: center;
`;

const PlayerName = styled.span`
  display: block;
  max-width: 100px;
  font-size: ${SizeMap.FS_20};
  line-height: 1.2;
  word-break: break-word;
`;

const Score = styled.span`
  display: block;
  font-size: ${SizeMap.FS_24};
`;
