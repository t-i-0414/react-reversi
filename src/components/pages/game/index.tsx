import React from 'react';
import styled from 'styled-components';
import ResetModal from '~/components/molecules/reset-modal';
import Board from '~/components/organisms/board';
import Score from '~/components/organisms/score';
import SettingForm from '~/components/organisms/setting-form';
import { ColorMap } from '~/const';
import useGame from './hooks';

export interface GameProps {
  dataCy?: string;
}
const Game: React.FC<GameProps> = ({ dataCy }) => {
  const { isGameStarted, isGameFinished } = useGame();

  return (
    <>
      <StyledHeading>React Reversi</StyledHeading>

      {isGameStarted && (
        <>
          <GameWrapper data-cy={dataCy}>
            <Board dataCy='board' />
            <Score />
          </GameWrapper>
          {isGameFinished && <ResetModal />}
        </>
      )}

      {!isGameStarted && <SettingForm dataCy={dataCy} />}
    </>
  );
};

const StyledHeading = styled.h1`
  margin: 12px 0;
  color: ${ColorMap.TX_DEEP_BLACK};
  text-align: center;
`;

const GameWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 1026px;
  height: 760px;
  margin: 0 auto;
  color: ${ColorMap.TX_BLACK};
`;

export default Game;
