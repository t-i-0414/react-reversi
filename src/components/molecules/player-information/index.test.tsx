import { expect } from '@jest/globals';
import React from 'react';
import renderer from 'react-test-renderer';
import { initialState } from '~/redux/modules/game';
import PlayerInformation from '.';

describe('playerInformation unit test', () => {
  it('snapshot(PieceColor White)', () => {
    expect.hasAssertions();

    const component = renderer.create(
      <PlayerInformation gamePlayer={initialState.players.black} />,
    );
    expect(component).toMatchSnapshot();
  });
  it('snapshot(PieceColor Black)', () => {
    expect.hasAssertions();

    const component = renderer.create(
      <PlayerInformation gamePlayer={initialState.players.white} />,
    );
    expect(component).toMatchSnapshot();
  });
});
