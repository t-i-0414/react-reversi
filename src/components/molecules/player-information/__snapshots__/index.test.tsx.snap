// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`playerInformation unit test snapshot(PieceColor Black) 1`] = `
<div
  className="sc-guDLRT dSWJiq"
>
  <span
    className="sc-jTQDnj dhbzrs"
  >
    0
  </span>
  <div
    className="sc-dmyDGi ZixER"
  >
    <span
      className="sc-beySbM bGJJsX"
      color="#fff"
    />
    <p
      className="sc-eDLJxc hmNxCF"
      color="#555"
    >
      P2
    </p>
  </div>
</div>
`;

exports[`playerInformation unit test snapshot(PieceColor White) 1`] = `
<div
  className="sc-guDLRT dSWJiq"
>
  <div
    className="sc-dmyDGi ZixER"
  >
    <span
      className="sc-beySbM ejsooN"
      color="#000"
    />
    <p
      className="sc-eDLJxc cTKOlq"
      color="#fff"
    >
      P1
    </p>
  </div>
  <span
    className="sc-jTQDnj dhbzrs"
  >
    0
  </span>
</div>
`;
